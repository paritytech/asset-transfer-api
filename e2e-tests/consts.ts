// Copyright 2023 Parity Technologies (UK) Ltd.

export const KUSAMA_ASSET_HUB_WS_URL = 'ws://127.0.0.1:9911';
export const ROCOCO_ALICE_WS_URL = 'ws://127.0.0.1:9900';
export const TRAPPIST_WS_URL = 'ws://127.0.0.1:9921';

export const BOB_ADDR = '5FHneW46xGXgs5mUiveU4sbTyGBzmstUspZC92UhjJM694ty';
export const FERDE_ADDR = '5CiPPseXPECbkjWCa6MnjNokrgYjMqmKndv2rSnekmSK2DjL';

export const EXTRINSIC_VERSION = 4;
