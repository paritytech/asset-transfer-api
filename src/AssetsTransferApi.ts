// Copyright 2023 Parity Technologies (UK) Ltd.

import '@polkadot/api-augment';

import type { ApiPromise } from '@polkadot/api';
import type { SubmittableExtrinsic } from '@polkadot/api/submittable/types';
import type { Option, u32 } from '@polkadot/types';
import type { ISubmittableResult } from '@polkadot/types/types';

import {
	DEFAULT_XCM_VERSION,
	SYSTEM_PARACHAINS_IDS,
	SYSTEM_PARACHAINS_NAMES,
} from './consts';
import { transfer, transferKeepAlive } from './createCalls';
import {
	limitedReserveTransferAssets,
	reserveTransferAssets,
} from './createXcmCalls';
import { establishXcmPallet } from './createXcmCalls/util/establishXcmPallet';
import { checkLocalTxInput, checkXcmVersion } from './errors';
import { parseRegistry } from './registry/parseRegistry';
import type { ChainInfoRegistry } from './registry/types';
import { sanitizeAddress } from './sanitize/sanitizeAddress';
import {
	ConstructedFormat,
	Format,
	IAssetsTransferApiOpts,
	IChainInfo,
	IDirection,
	IMethods,
	ITransferArgsOpts,
	TxResult,
} from './types';

export class AssetsTransferApi {
	readonly _api: ApiPromise;
	readonly _opts: IAssetsTransferApiOpts;
	readonly _info: Promise<IChainInfo>;
	readonly _safeXcmVersion: Promise<u32>;
	readonly _registry: ChainInfoRegistry;

	constructor(api: ApiPromise, opts: IAssetsTransferApiOpts = {}) {
		this._api = api;
		this._opts = opts;
		this._info = this.fetchChainInfo();
		this._safeXcmVersion = this.fetchSafeXcmVersion();
		this._registry = parseRegistry(opts);
	}

	/**
	 * Create an asset transfer transaction. This can be either locally on a systems parachain,
	 * or between chains using xcm.
	 *
	 * @param destChainId ID of the destination (para) chain (‘0’ for Relaychain)
	 * @param destAddr Address of destination account
	 * @param assetIds Array of assetId's to be transferred (‘0’ for Native Relay Token)
	 * @param amounts Array of the amounts of each token to transfer
	 * @param opts Options
	 */
	public async createTransferTransaction<T extends Format>(
		destChainId: string,
		destAddr: string,
		assetIds: string[],
		amounts: string[],
		opts?: ITransferArgsOpts<T>
	): Promise<TxResult<T>> {
		const { _api, _info, _safeXcmVersion } = this;
		const { specName } = await _info;
		const safeXcmVersion = await _safeXcmVersion;
		const isOriginSystemParachain = SYSTEM_PARACHAINS_NAMES.includes(
			specName.toLowerCase()
		);
		const isDestSystemParachain = SYSTEM_PARACHAINS_IDS.includes(destChainId);

		/**
		 * Sanitize the address to a hex, and ensure that the passed in SS58, or publickey
		 * is validated correctly.
		 */
		const addr = sanitizeAddress(destAddr);
		/**
		 * Create a local asset transfer.
		 */
		if (isDestSystemParachain && isOriginSystemParachain) {
			/**
			 * This will throw a BaseError if the inputs are incorrect and don't
			 * fit the constraints for creating a local asset transfer.
			 */
			checkLocalTxInput(assetIds, amounts); // Throws an error when any of the inputs are incorrect.
			const method = opts?.keepAlive ? 'transferKeepAlive' : 'transfer';
			const tx =
				method === 'transferKeepAlive'
					? transferKeepAlive(_api, addr, assetIds[0], amounts[0])
					: transfer(_api, addr, assetIds[0], amounts[0]);

			return this.constructFormat(tx, 'local', null, method, opts?.format);
		}

		/**
		 * Establish the Transaction Direction
		 */
		const xcmDirection = this.establishDirection(destChainId, specName);
		const isRelayDirection = xcmDirection.toLowerCase().includes('relay');
		/**
		 * Establish an xcmVersion
		 */
		const xcmVersion =
			opts?.xcmVersion === undefined
				? safeXcmVersion.toNumber()
				: opts.xcmVersion;
		checkXcmVersion(xcmVersion); // Throws an error when the xcmVersion is not supported.

		// TODO: Check for xcm construction errors depending on the input.

		/**
		 * Lengths should match, and indicies between both the amounts and assetIds should match.
		 */
		if (assetIds.length !== amounts.length && !isRelayDirection) {
			throw Error(
				'`amounts`, and `assetIds` fields should match in length when constructing a tx from a parachain to a parachain or locally on a system parachain.'
			);
		}

		let txMethod: IMethods;
		let transaction: SubmittableExtrinsic<'promise', ISubmittableResult>;
		if (opts?.isLimited) {
			txMethod = 'limitedReserveTransferAssets';
			transaction = limitedReserveTransferAssets(
				_api,
				xcmDirection,
				addr,
				assetIds,
				amounts,
				destChainId,
				xcmVersion,
				opts?.weightLimit
			);
		} else {
			txMethod = 'reserveTransferAssets';
			transaction = reserveTransferAssets(
				_api,
				xcmDirection,
				addr,
				assetIds,
				amounts,
				destChainId,
				xcmVersion
			);
		}

		return this.constructFormat<T>(
			transaction,
			xcmDirection,
			xcmVersion,
			txMethod,
			opts?.format
		);
	}

	/**
	 * Fetch runtime information based on the connected chain.
	 *
	 * @param api ApiPromise
	 */
	private async fetchChainInfo(): Promise<IChainInfo> {
		const { _api } = this;
		const { specName, specVersion } = await _api.rpc.state.getRuntimeVersion();
		return {
			specName: specName.toString(),
			specVersion: specVersion.toString(),
		};
	}

	/**
	 * Declare the direction of the xcm message.
	 *
	 * @param destChainId
	 * @param specName
	 */
	private establishDirection(
		destChainId: string,
		specName: string
	): IDirection {
		const { _api } = this;
		const isSystemParachain = SYSTEM_PARACHAINS_NAMES.includes(
			specName.toLowerCase()
		);
		const isDestIdSystemPara = SYSTEM_PARACHAINS_IDS.includes(destChainId);

		/**
		 * Check if the origin is a System Parachain
		 */
		if (isSystemParachain && destChainId === '0') {
			return IDirection.SystemToRelay;
		}

		if (isSystemParachain && destChainId !== '0') {
			return IDirection.SystemToPara;
		}

		/**
		 * Check if the origin is a Relay Chain
		 */
		if (_api.query.paras && isDestIdSystemPara) {
			return IDirection.RelayToSystem;
		}

		if (_api.query.paras && !isDestIdSystemPara) {
			return IDirection.RelayToPara;
		}

		/**
		 * Check if the origin is a Parachain or Parathread
		 */
		if (_api.query.polkadotXcm && !isDestIdSystemPara) {
			throw Error('ParaToRelay is not yet implemented');

			return IDirection.ParaToRelay;
		}

		if (_api.query.polkadotXcm) {
			throw Error('ParaToPara is not yet implemented');

			return IDirection.ParaToPara;
		}

		throw Error('Could not establish a xcm transaction direction');
	}

	/**
	 * Construct the correct format for the transaction.
	 * If nothing is passed in, the format will default to a signing payload.
	 *
	 * @param tx A polkadot-js submittable extrinsic
	 * @param format The format to return the tx in.
	 */
	private constructFormat<T extends Format>(
		tx: SubmittableExtrinsic<'promise', ISubmittableResult>,
		direction: IDirection | 'local',
		xcmVersion: number | null = null,
		method: IMethods,
		format?: T
	): TxResult<T> {
		const { _api } = this;
		const fmt = format ? format : 'payload';
		const result: TxResult<T> = {
			direction,
			xcmVersion,
			method,
			format: fmt as Format | 'local',
			tx: '' as ConstructedFormat<T>,
		};

		if (fmt === 'call') {
			result.tx = _api.registry
				.createType('Call', {
					callIndex: tx.callIndex,
					args: tx.args,
				})
				.toHex() as ConstructedFormat<T>;
		}

		if (fmt === 'submittable') {
			result.tx = tx as ConstructedFormat<T>;
		}

		if (fmt === 'payload') {
			result.tx = _api.registry
				.createType('ExtrinsicPayload', tx, {
					version: tx.version,
				})
				.toHex() as ConstructedFormat<T>;
		}

		return result;
	}

	/**
	 * Fetch for a safe Xcm Version from the chain, if none exists the
	 * in app default version will be used.
	 */
	private async fetchSafeXcmVersion(): Promise<u32> {
		const { _api } = this;
		const pallet = establishXcmPallet(_api);
		const safeVersion = await _api.query[pallet].safeXcmVersion<Option<u32>>();
		const version = safeVersion.isSome
			? safeVersion.unwrap()
			: _api.registry.createType('u32', DEFAULT_XCM_VERSION);

		return version;
	}
}
