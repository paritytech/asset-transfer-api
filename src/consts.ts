// Copyright 2023 Parity Technologies (UK) Ltd.

/**
 * List of all known relay chains.
 */
export const RELAY_CHAIN_NAMES = ['kusama', 'polkadot', 'westend', 'rococo'];

/**
 * As of now all the known relay chains have an ID of 0.
 */
export const RELAY_CHAIN_IDS = ['0'];

/**
 * AssetHub chains have an ID of 1000;
 */
export const ASSET_HUB_CHAIN_ID = '1000';

/**
 * List of all known system parachains.
 */
export const SYSTEM_PARACHAINS_NAMES = [
	'statemine',
	'statemint',
	'westmint',
	'asset-hub-kusama',
	'asset-hub-polkadot',
	'asset-hub-westend',
	'asset-hub-rococo',
	'bridge-hub-kusama',
	'bridge-hub-polkadot',
	'encointer-parachain',
	'collectives',
];
export const POLKADOT_ASSET_HUB_SPEC_NAMES = ['statemint', 'asset-hub-polkadot'];
export const KUSAMA_ASSET_HUB_SPEC_NAMES = ['statemine', 'asset-hub-kusama'];
export const WESTEND_ASSET_HUB_SPEC_NAMES = ['westmint', 'asset-hub-westend'];
export const ROCOCO_ASSET_HUB_SPEC_NAME = ['asset-hub-rococo'];

/**
 * The default xcm version to construct a xcm message.
 */
export const DEFAULT_XCM_VERSION = 2;
/**
 * There should only ever be two supported versions.
 * Therefore supported xcm versions should have a fixed length of 2
 */
export const SUPPORTED_XCM_VERSIONS: [number, number] = [2, 3];
/**
 * The current maximum number of assets that can be transferred in an extrinsic
 * https://github.com/paritytech/polkadot/blob/e0ed7e862c8c8b6c75eda1731c449543642176ef/xcm/pallet-xcm/src/lib.rs#L1131
 */
export const MAX_ASSETS_FOR_TRANSFER = 2;

/**
 * This is the max length for a number before we need to use BigInt.
 */
export const MAX_NUM_LENGTH = Number.MAX_SAFE_INTEGER.toString().length;

/**
 * URL to reach the CDN endpoint.
 */
export const CDN_URL = 'https://paritytech.github.io/asset-transfer-api-registry/registry.json';
