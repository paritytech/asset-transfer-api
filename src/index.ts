// Copyright 2024 Parity Technologies (UK) Ltd.

export * from './AssetTransferApi';
export * from './constructApiPromise';
export * from './types';
