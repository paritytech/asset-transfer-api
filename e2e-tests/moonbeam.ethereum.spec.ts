import { setupNetworks, testingPairs, withExpect } from '@acala-network/chopsticks-testing';
import { NetworkContext } from '@acala-network/chopsticks-utils';
import { afterEach, beforeEach, expect, test } from 'vitest';

import { AssetTransferApi } from '../src/AssetTransferApi';
import { ETHEREUM_MAINNET_NETWORK_GLOBAL_CONSENSUS_LOCATION } from '../src/consts';

const { checkSystemEvents } = withExpect(expect);

describe('Moonbeam <> Ethereum', () => {
	let moonbeam: NetworkContext;
	let polkadotAssetHub: NetworkContext;
	let polkadotBridgeHub: NetworkContext;

	const { alice, alith } = testingPairs();

	const moonbeamPort = 8016;
	const polkadotBridgeHubPort = 8017;
	const polkadotAssetHubPort = 8018;
	const runtimeLogLevel = 0;

	beforeEach(async () => {
		const { moonbeam1, polkadotBridgeHub1, polkadotAssetHub1 } = await setupNetworks({
			moonbeam1: {
				endpoint: 'wss://moonbeam.public.blastapi.io',
				db: `./chopsticks-db/db.sqlite-moonbeam-${moonbeamPort}`,
				port: moonbeamPort,
				runtimeLogLevel,
			},
			polkadotBridgeHub1: {
				endpoint: 'wss://polkadot-bridge-hub-rpc.polkadot.io',
				db: `./chopsticks-db/db.sqlite-polkadot-bridge-hub-${polkadotBridgeHubPort}`,
				port: polkadotBridgeHubPort,
				runtimeLogLevel,
			},
			polkadotAssetHub1: {
				endpoint: 'wss://polkadot-asset-hub-rpc.polkadot.io',
				db: `./chopsticks-db/db.sqlite-polkadot-asset-hub-${polkadotAssetHubPort}`,
				port: polkadotAssetHubPort,
				runtimeLogLevel,
			},
		});

		moonbeam = moonbeam1;
		polkadotBridgeHub = polkadotBridgeHub1;
		polkadotAssetHub = polkadotAssetHub1;
	}, 1000000);

	afterEach(async () => {
		await moonbeam.teardown();
		await polkadotAssetHub.teardown();
		await polkadotBridgeHub.teardown();
	}, 1000000);

	describe('XCM V3', () => {
		const xcmVersion = 3;

		test('Transfer Snowbridge WETH from Moonbeam to Ethereum', async () => {
			await moonbeam.dev.setStorage({
				System: {
					Account: [
						[[alith.address], { providers: 1, data: { free: '100000000000000000000000' } }], // GLMR
					],
				},
				Assets: {
					Account: [
						[['42259045809535163221576417993425387648', alith.address], { balance: '1000000000000000' }], // DOT
						[['178794693648360392906933130845919698647', alith.address], { balance: '1000000000000000' }], // Snowbridge WETH
					],
				},
			});

			const assetTransferApi = new AssetTransferApi(moonbeam.api, 'moonbeam', xcmVersion, {
				registryType: 'NPM',
				injectedRegistry: {
					polkadot: {
						2004: {
							tokens: [],
							assetsInfo: {},
							foreignAssetsInfo: {},
							poolPairsInfo: {},
							specName: 'moonbeam',
							xcAssetsData: [
								{
									paraID: 0,
									symbol: 'WETH.snow',
									decimals: 18,
									xcmV1MultiLocation:
										'{"v1":{"parents":2,"interior":{"x2":[{"globalConsensus":{"ethereum":{"chainId":1}}},{"accountKey20":{"network":null,"key":"0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2"}}]}}}',
									asset: '178794693648360392906933130845919698647',
									assetHubReserveLocation: '{"parents":"1","interior":{"X1":{"Parachain":"1000"}}}',
								},
							],
						},
					},
				},
			});

			const tx = await assetTransferApi.createTransferTransaction(
				ETHEREUM_MAINNET_NETWORK_GLOBAL_CONSENSUS_LOCATION,
				alith.address,
				['DOT', 'WETH.snow'],
				['100000000000', '750000000000000'],
				{
					sendersAddr: alice.address,
					format: 'payload',
					xcmVersion,
					paysWithFeeDest: 'DOT',
				},
			);

			const extrinsic = assetTransferApi.api.registry.createType('Extrinsic', { method: tx.tx.method }, { version: 4 });

			await moonbeam.api.tx(extrinsic).signAndSend(alith);
			await moonbeam.dev.newBlock();

			await checkSystemEvents(moonbeam, 'polkadotXcm')
				.redact({ redactKeys: new RegExp('messageId|proofSize|refTime') })
				.toMatchSnapshot('Moonbeam xcm message sent');

			const polkadotAssetHubCurrentChainHead = polkadotAssetHub.chain.head.number;
			await polkadotAssetHub.dev.newBlock();
			await polkadotAssetHub.dev.setHead(polkadotAssetHubCurrentChainHead + 1);

			await checkSystemEvents(polkadotAssetHub, 'foreignAssets').toMatchSnapshot('assethub foreign assets burned');
			await checkSystemEvents(polkadotAssetHub, 'xcmpQueue', 'XcmpMessageSent').toMatchSnapshot(
				'assetHub xcmp message sent',
			);

			const polkadotBridgeHubCurrentChainHead = polkadotBridgeHub.chain.head.number;
			await polkadotBridgeHub.dev.newBlock();
			await polkadotBridgeHub.dev.setHead(polkadotBridgeHubCurrentChainHead + 1);

			await checkSystemEvents(polkadotBridgeHub, 'ethereumOutboundQueue')
				.redact({ redactKeys: new RegExp('nonce') })
				.toMatchSnapshot('bridgehub ethereum outbound queue events');
		}, 200000);
	});
	describe('XCM V4', () => {
		const xcmVersion = 4;

		test('Transfer Snowbridge WETH from Moonbeam to Ethereum', async () => {
			await moonbeam.dev.setStorage({
				System: {
					Account: [
						[[alith.address], { providers: 1, data: { free: '100000000000000000000000' } }], // GLMR
					],
				},
				Assets: {
					Account: [
						[['42259045809535163221576417993425387648', alith.address], { balance: '1000000000000000' }], // DOT
						[['178794693648360392906933130845919698647', alith.address], { balance: '1000000000000000' }], // Snowbridge WETH
					],
				},
			});

			const assetTransferApi = new AssetTransferApi(moonbeam.api, 'moonbeam', xcmVersion, {
				registryType: 'NPM',
				injectedRegistry: {
					polkadot: {
						2030: {
							tokens: [],
							assetsInfo: {},
							foreignAssetsInfo: {},
							poolPairsInfo: {},
							specName: 'moonbeam',
							xcAssetsData: [
								{
									paraID: 0,
									symbol: 'WETH.snow',
									decimals: 18,
									xcmV1MultiLocation:
										'{"v1":{"parents":2,"interior":{"x2":[{"globalConsensus":{"ethereum":{"chainId":1}}},{"accountKey20":{"network":null,"key":"0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2"}}]}}}',
									asset: { Token2: '13' },
									assetHubReserveLocation: '{"parents":"1","interior":{"X1":{"Parachain":"1000"}}}',
								},
							],
						},
					},
				},
			});

			const tx = await assetTransferApi.createTransferTransaction(
				ETHEREUM_MAINNET_NETWORK_GLOBAL_CONSENSUS_LOCATION,
				alith.address,
				['DOT', 'WETH.snow'],
				['100000000000', '750000000000000'],
				{
					sendersAddr: alice.address,
					format: 'payload',
					xcmVersion,
					paysWithFeeDest: 'DOT',
				},
			);

			const extrinsic = assetTransferApi.api.registry.createType('Extrinsic', { method: tx.tx.method }, { version: 4 });

			await moonbeam.api.tx(extrinsic).signAndSend(alith);
			await moonbeam.dev.newBlock();

			await checkSystemEvents(moonbeam, 'polkadotXcm')
				.redact({ redactKeys: new RegExp('messageId|proofSize|refTime') })
				.toMatchSnapshot('Moonbeam xcm message sent');

			const polkadotAssetHubCurrentChainHead = polkadotAssetHub.chain.head.number;
			await polkadotAssetHub.dev.newBlock();
			await polkadotAssetHub.dev.setHead(polkadotAssetHubCurrentChainHead + 1);

			await checkSystemEvents(polkadotAssetHub, 'foreignAssets').toMatchSnapshot('assethub foreign assets burned');
			await checkSystemEvents(polkadotAssetHub, 'xcmpQueue', 'XcmpMessageSent').toMatchSnapshot(
				'assetHub xcmp message sent',
			);

			const assetHubEvents = await polkadotAssetHub.api.query.system.events();

			const xcmMessageProcessed = assetHubEvents[assetHubEvents.length - 1];
			expect(xcmMessageProcessed.phase.toString()).toEqual('Finalization');
			expect(xcmMessageProcessed.event.method).toEqual('Processed');
			expect(xcmMessageProcessed.event.section).toEqual('messageQueue');

			const polkadotBridgeHubCurrentChainHead = polkadotBridgeHub.chain.head.number;
			await polkadotBridgeHub.dev.newBlock();
			await polkadotBridgeHub.dev.setHead(polkadotBridgeHubCurrentChainHead + 1);

			await checkSystemEvents(polkadotBridgeHub, 'ethereumOutboundQueue')
				.redact({ redactKeys: new RegExp('nonce') })
				.toMatchSnapshot('bridgehub ethereum outbound queue events');

			const bridgeHubEvents = await polkadotBridgeHub.api.query.system.events();
			const messageAcceptedEvent = bridgeHubEvents[bridgeHubEvents.length - 3];
			expect(messageAcceptedEvent.event.section).toEqual('ethereumOutboundQueue');
			expect(messageAcceptedEvent.event.method).toEqual('MessageAccepted');

			const messageCommittedEvent = bridgeHubEvents[bridgeHubEvents.length - 1];
			expect(messageCommittedEvent.event.section).toEqual('ethereumOutboundQueue');
			expect(messageCommittedEvent.event.method).toEqual('MessagesCommitted');
		}, 200000);
	});
});
