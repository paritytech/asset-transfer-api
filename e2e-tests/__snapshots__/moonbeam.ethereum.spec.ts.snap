// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Moonbeam <> Ethereum > XCM V3 > Transfer Snowbridge WETH from Moonbeam to Ethereum > Moonbeam xcm message sent 1`] = `
[
  {
    "data": {
      "outcome": {
        "Complete": {
          "used": {
            "proofSize": "(rounded 19000)",
            "refTime": "(rounded 7700000000)",
          },
        },
      },
    },
    "method": "Attempted",
    "section": "polkadotXcm",
  },
  {
    "data": {
      "fees": [],
      "paying": {
        "interior": {
          "X1": [
            {
              "AccountKey20": {
                "key": "0xf24ff3a9cf04c71dbc94d0b566f7a27b94566cac",
                "network": "Polkadot",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "FeesPaid",
    "section": "polkadotXcm",
  },
  {
    "data": {
      "destination": {
        "interior": {
          "X1": [
            {
              "Parachain": 1000,
            },
          ],
        },
        "parents": 1,
      },
      "message": [
        {
          "WithdrawAsset": [
            {
              "fun": {
                "Fungible": 100000000000,
              },
              "id": {
                "interior": "Here",
                "parents": 1,
              },
            },
            {
              "fun": {
                "Fungible": 750000000000000,
              },
              "id": {
                "interior": {
                  "X2": [
                    {
                      "GlobalConsensus": {
                        "Ethereum": {
                          "chainId": 1,
                        },
                      },
                    },
                    {
                      "AccountKey20": {
                        "key": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                        "network": null,
                      },
                    },
                  ],
                },
                "parents": 2,
              },
            },
          ],
        },
        "ClearOrigin",
        {
          "BuyExecution": {
            "fees": {
              "fun": {
                "Fungible": 100000000000,
              },
              "id": {
                "interior": "Here",
                "parents": 1,
              },
            },
            "weightLimit": "Unlimited",
          },
        },
        {
          "SetAppendix": [
            {
              "DepositAsset": {
                "assets": {
                  "Wild": "All",
                },
                "beneficiary": {
                  "interior": {
                    "X1": [
                      {
                        "AccountId32": {
                          "id": "(hash)",
                          "network": null,
                        },
                      },
                    ],
                  },
                  "parents": 0,
                },
              },
            },
          ],
        },
        {
          "InitiateReserveWithdraw": {
            "assets": {
              "Wild": {
                "AllOf": {
                  "fun": "Fungible",
                  "id": {
                    "interior": {
                      "X2": [
                        {
                          "GlobalConsensus": {
                            "Ethereum": {
                              "chainId": 1,
                            },
                          },
                        },
                        {
                          "AccountKey20": {
                            "key": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                            "network": null,
                          },
                        },
                      ],
                    },
                    "parents": 2,
                  },
                },
              },
            },
            "reserve": {
              "interior": {
                "X1": [
                  {
                    "GlobalConsensus": {
                      "Ethereum": {
                        "chainId": 1,
                      },
                    },
                  },
                ],
              },
              "parents": 2,
            },
            "xcm": [
              {
                "BuyExecution": {
                  "fees": {
                    "fun": {
                      "Fungible": 1,
                    },
                    "id": {
                      "interior": {
                        "X1": [
                          {
                            "AccountKey20": {
                              "key": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                              "network": null,
                            },
                          },
                        ],
                      },
                      "parents": 0,
                    },
                  },
                  "weightLimit": "Unlimited",
                },
              },
              {
                "DepositAsset": {
                  "assets": {
                    "Wild": {
                      "AllCounted": 1,
                    },
                  },
                  "beneficiary": {
                    "interior": {
                      "X1": [
                        {
                          "AccountKey20": {
                            "key": "0xf24ff3a9cf04c71dbc94d0b566f7a27b94566cac",
                            "network": null,
                          },
                        },
                      ],
                    },
                    "parents": 0,
                  },
                },
              },
              {
                "SetTopic": 0,
              },
            ],
          },
        },
        {
          "SetTopic": 0,
        },
      ],
      "messageId": "(redacted)",
      "origin": {
        "interior": {
          "X1": [
            {
              "AccountKey20": {
                "key": "0xf24ff3a9cf04c71dbc94d0b566f7a27b94566cac",
                "network": "Polkadot",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "Sent",
    "section": "polkadotXcm",
  },
]
`;

exports[`Moonbeam <> Ethereum > XCM V3 > Transfer Snowbridge WETH from Moonbeam to Ethereum > assetHub xcmp message sent 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
    },
    "method": "XcmpMessageSent",
    "section": "xcmpQueue",
  },
]
`;

exports[`Moonbeam <> Ethereum > XCM V3 > Transfer Snowbridge WETH from Moonbeam to Ethereum > assethub foreign assets burned 1`] = `
[
  {
    "data": {
      "assetId": {
        "interior": {
          "X2": [
            {
              "GlobalConsensus": {
                "Ethereum": {
                  "chainId": 1,
                },
              },
            },
            {
              "AccountKey20": {
                "key": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                "network": null,
              },
            },
          ],
        },
        "parents": 2,
      },
      "balance": 750000000000000,
      "owner": "13cKp89NgPL56sRoVRpBcjkGZPrk4Vf4tS6ePUD96XhAXozG",
    },
    "method": "Burned",
    "section": "foreignAssets",
  },
]
`;

exports[`Moonbeam <> Ethereum > XCM V3 > Transfer Snowbridge WETH from Moonbeam to Ethereum > bridgehub ethereum outbound queue events 1`] = `
[
  {
    "data": {
      "id": 0,
    },
    "method": "MessageQueued",
    "section": "ethereumOutboundQueue",
  },
  {
    "data": {
      "id": 0,
      "nonce": "(redacted)",
    },
    "method": "MessageAccepted",
    "section": "ethereumOutboundQueue",
  },
  {
    "data": {
      "count": 1,
      "root": "(hash)",
    },
    "method": "MessagesCommitted",
    "section": "ethereumOutboundQueue",
  },
]
`;

exports[`Moonbeam <> Ethereum > XCM V4 > Transfer Snowbridge WETH from Moonbeam to Ethereum > Moonbeam xcm message sent 1`] = `
[
  {
    "data": {
      "outcome": {
        "Complete": {
          "used": {
            "proofSize": "(rounded 19000)",
            "refTime": "(rounded 7700000000)",
          },
        },
      },
    },
    "method": "Attempted",
    "section": "polkadotXcm",
  },
  {
    "data": {
      "fees": [],
      "paying": {
        "interior": {
          "X1": [
            {
              "AccountKey20": {
                "key": "0xf24ff3a9cf04c71dbc94d0b566f7a27b94566cac",
                "network": "Polkadot",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "FeesPaid",
    "section": "polkadotXcm",
  },
  {
    "data": {
      "destination": {
        "interior": {
          "X1": [
            {
              "Parachain": 1000,
            },
          ],
        },
        "parents": 1,
      },
      "message": [
        {
          "WithdrawAsset": [
            {
              "fun": {
                "Fungible": 100000000000,
              },
              "id": {
                "interior": "Here",
                "parents": 1,
              },
            },
            {
              "fun": {
                "Fungible": 750000000000000,
              },
              "id": {
                "interior": {
                  "X2": [
                    {
                      "GlobalConsensus": {
                        "Ethereum": {
                          "chainId": 1,
                        },
                      },
                    },
                    {
                      "AccountKey20": {
                        "key": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                        "network": null,
                      },
                    },
                  ],
                },
                "parents": 2,
              },
            },
          ],
        },
        "ClearOrigin",
        {
          "BuyExecution": {
            "fees": {
              "fun": {
                "Fungible": 100000000000,
              },
              "id": {
                "interior": "Here",
                "parents": 1,
              },
            },
            "weightLimit": "Unlimited",
          },
        },
        {
          "SetAppendix": [
            {
              "DepositAsset": {
                "assets": {
                  "Wild": "All",
                },
                "beneficiary": {
                  "interior": {
                    "X1": [
                      {
                        "AccountId32": {
                          "id": "(hash)",
                          "network": null,
                        },
                      },
                    ],
                  },
                  "parents": 0,
                },
              },
            },
          ],
        },
        {
          "InitiateReserveWithdraw": {
            "assets": {
              "Wild": {
                "AllOf": {
                  "fun": "Fungible",
                  "id": {
                    "interior": {
                      "X2": [
                        {
                          "GlobalConsensus": {
                            "Ethereum": {
                              "chainId": 1,
                            },
                          },
                        },
                        {
                          "AccountKey20": {
                            "key": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                            "network": null,
                          },
                        },
                      ],
                    },
                    "parents": 2,
                  },
                },
              },
            },
            "reserve": {
              "interior": {
                "X1": [
                  {
                    "GlobalConsensus": {
                      "Ethereum": {
                        "chainId": 1,
                      },
                    },
                  },
                ],
              },
              "parents": 2,
            },
            "xcm": [
              {
                "BuyExecution": {
                  "fees": {
                    "fun": {
                      "Fungible": 1,
                    },
                    "id": {
                      "interior": {
                        "X1": [
                          {
                            "AccountKey20": {
                              "key": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                              "network": null,
                            },
                          },
                        ],
                      },
                      "parents": 0,
                    },
                  },
                  "weightLimit": "Unlimited",
                },
              },
              {
                "DepositAsset": {
                  "assets": {
                    "Wild": {
                      "AllCounted": 1,
                    },
                  },
                  "beneficiary": {
                    "interior": {
                      "X1": [
                        {
                          "AccountKey20": {
                            "key": "0xf24ff3a9cf04c71dbc94d0b566f7a27b94566cac",
                            "network": null,
                          },
                        },
                      ],
                    },
                    "parents": 0,
                  },
                },
              },
              {
                "SetTopic": 0,
              },
            ],
          },
        },
        {
          "SetTopic": 0,
        },
      ],
      "messageId": "(redacted)",
      "origin": {
        "interior": {
          "X1": [
            {
              "AccountKey20": {
                "key": "0xf24ff3a9cf04c71dbc94d0b566f7a27b94566cac",
                "network": "Polkadot",
              },
            },
          ],
        },
        "parents": 0,
      },
    },
    "method": "Sent",
    "section": "polkadotXcm",
  },
]
`;

exports[`Moonbeam <> Ethereum > XCM V4 > Transfer Snowbridge WETH from Moonbeam to Ethereum > assetHub xcmp message sent 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
    },
    "method": "XcmpMessageSent",
    "section": "xcmpQueue",
  },
]
`;

exports[`Moonbeam <> Ethereum > XCM V4 > Transfer Snowbridge WETH from Moonbeam to Ethereum > assethub foreign assets burned 1`] = `
[
  {
    "data": {
      "assetId": {
        "interior": {
          "X2": [
            {
              "GlobalConsensus": {
                "Ethereum": {
                  "chainId": 1,
                },
              },
            },
            {
              "AccountKey20": {
                "key": "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                "network": null,
              },
            },
          ],
        },
        "parents": 2,
      },
      "balance": 750000000000000,
      "owner": "13cKp89NgPL56sRoVRpBcjkGZPrk4Vf4tS6ePUD96XhAXozG",
    },
    "method": "Burned",
    "section": "foreignAssets",
  },
]
`;

exports[`Moonbeam <> Ethereum > XCM V4 > Transfer Snowbridge WETH from Moonbeam to Ethereum > bridgehub ethereum outbound queue events 1`] = `
[
  {
    "data": {
      "id": 0,
    },
    "method": "MessageQueued",
    "section": "ethereumOutboundQueue",
  },
  {
    "data": {
      "id": 0,
      "nonce": "(redacted)",
    },
    "method": "MessageAccepted",
    "section": "ethereumOutboundQueue",
  },
  {
    "data": {
      "count": 1,
      "root": "(hash)",
    },
    "method": "MessagesCommitted",
    "section": "ethereumOutboundQueue",
  },
]
`;
